File: HIndex.cpp
//Time Complexity O(n)
// Space Complexity O(n)
//Problem successfully executed on leetcode
//No problems faced while coading this.


#include<iostream>
#include<vector>
#include<queue>
#include<map>
#include<stack>
using namespace std;


class Solution {
public:
    int hIndex(vector<int>& citations) {
        int n=citations.size();
        int mid;
        int low=0;
        int high=n-1;
        while(high>=low)
        {
            mid = low + (high-low)/2;
            int midElement=citations[mid];
            int midMore=n-mid;
            
            if(midElement==midMore)
            {
                return midMore;
            }
            else if(midElement>midMore)
            {
                high=mid-1;
            }
            else
            {
                low=mid+1;
            }
            
        }
        return n-low;
    }
};

File: h-index.py
# TC O(log n)
# SC O(1)
class Solution:
    def hIndex(self, citations: List[int]) -> int:
        # diff = n - mid == citations[mid] : return n - mid
        # if n - mid > citations[mid] low = mid + 1
        # else high = mid - 1
        n = len(citations)

        low, high = 0, n - 1
        while low <= high:
            mid = low + (high-low) // 2
            diff = n - mid
            if diff == citations[mid]:
                return diff
            elif diff > citations[mid]:
                low = mid + 1
            else:
                high = mid - 1

        return n - low
        